marimbabot_speech/src/speech2text.py

```diff
@@ -1,84 +1,109 @@
 import os
 from nvidia import cudnn
 os.environ['LD_LIBRARY_PATH'] = f'{os.path.dirname(cudnn.__file__)}/lib:$LD_LIBRARY_PATH'
-# since I haven't specified the version of TensorFlow, it seems to install the newest version, so according to official instruction, the environment needs to be configured.
-
+# since I haven't specified the version of TensorFlow, it seems to install the newest version,
+# so according to official instruction, the environment needs to be configured.
+from marimbabot_msgs.msg import Speech as SpeechMsg
+from audio_common_msgs.msg import AudioData
 import rospy
 import whisper
-from utils.file_control import WAVFile
-from marimbabot_msgs.msg import Speech as SpeechMsg
-from marimbabot_msgs.msg import TmpFile as TmpFileMsg
-import time
-from utils.command import get_commands
+import numpy as np
+import struct
 
+# Speech to text recognition, it is a wrapper of whisper
 class STT:
 	def __init__(self):
+		self.prompt = self.generate_prompt()
+		self.init_audio_config()
+		self.init_whisper_model()
+		self.warm_up()  # warm up the whisper model
+		self.init_ros_node()
+
+	def init_audio_config(self):
+		self.sr = rospy.get_param('~sr', 16000)
+
+	def init_whisper_model(self):
 		# 'tiny.en', 'tiny', 'base.en', 'base', 'small.en', 'small', 'medium.en', 'medium', 'large-v1', 'large-v2', 'large'
 		self.model = whisper.load_model('medium.en')
-		self.file_controller = WAVFile()
-		self.speech_pub = rospy.Publisher('/speech_node/speech', SpeechMsg, queue_size=100, tcp_nodelay=True)
-		self.tmp_sub = rospy.Subscriber('/speech_node/audio_tmp', TmpFileMsg, self.tmp_callback, queue_size=10, tcp_nodelay=True)
-		self.recognize_freq = 2  # Hz
-		self.recognize_rate = rospy.Rate(self.recognize_freq)
-		self.no_speech_prob_filter = 0.5
-		rospy.logdebug(f"cwd:{os.getcwd()}")
-		self.prompt = self.generate_prompt()
 
-	def generate_prompt(self):
-		commands = get_commands()
-		base_prompt = '''Marimbabot is a marimba playing robot arm. The possible commands include:		
-		'''
-		for command in commands:
-			base_prompt += ''.join(f'{command.strip()}, ')
-		rospy.logdebug(f"Generate prompt as: {base_prompt}.")
-		return base_prompt
+	def warm_up(self):
```

Comment:
Why is this needed? You pad to 30 sec ether way right?

Reply:
input 30sec blank data to warm up the whisper, since the whisper takes the fixed length of audio data as input.  Or we also can input random length of data, then we must pad or trim it to 30 sec.